pkg/monitoring/docker/docker.go

```diff
@@ -11,29 +12,47 @@ import (
 
 const dockerEndpointAddress = "unix:///var/run/docker.sock"
 
-var client *docker.Client
-var connectionSucceedOnce bool
+type DockerWatcher struct {
+	client                *docker.Client
+	connectionSucceedOnce bool
+}
 
-func ListContainers() (map[string]interface{}, error) {
-	var err error
-	if client == nil {
-		client, err = docker.NewClient(dockerEndpointAddress)
+func (dw *DockerWatcher) initClient() error {
+	if dw.client == nil {
+		_, err := docker.NewClient(dockerEndpointAddress)
 		if err != nil {
-			return nil, err
+			return err
 		}
 	}
+	return nil
+}
 
-	containers, err := client.ListContainers(docker.ListContainersOptions{All: false})
-	if err != nil && err == docker.ErrConnectionRefused && !connectionSucceedOnce {
-		// do not produce error if Docker is missing and wasn't successfully connected in this session before
+func (dw *DockerWatcher) ListContainers() (map[string]interface{}, error) {
+	err := dw.initClient()
+	if err != nil {
+		return nil, err
+	}
+
+	var dockerExecExists bool
+	if _, err := exec.LookPath("docker"); err == nil {
+		dockerExecExists = true
+	}
+
+	containers, err := dw.client.ListContainers(docker.ListContainersOptions{All: false})
+	if err != nil &&
+		!dockerExecExists && // docker executable not exists in the system
```

Comment:
Extraction of `DockerWatcher` 👍 

Can we also simplify the condition here? Maybe split into two `if`?

Reply:
I have added the comments. Let's leave it as is. I didn't find the way to split it elegantly

Reply:
At least clean up the `else if` then. This is a pattern I already mentioned many times.
If you have a `return` inside the `if` branch you _don't_ have to put an `else if` afterwards. Omit the `else` because if the first branch is triggered, the function/method returns anyway and therefor the `else` is not needed.

Please try to keep that in mind for the future.

Reply:
Right, `else` is not necessary here. But I think it improves the readability.
If `else` is omitted when someone the code fluently he can think that the second error handling is the only one

Reply:
> But I think it improves the readability in this case.

I disagree. As you said, it's not needed and therefor should be removed.
The first `if` checks for the specific error and the second `if` checks for the general error.
Add a comment to each `if` to describe what this check does, done.

> he can think that the second error handling is the only one

So you write the code for someone not reading the code properly?

Reply:
take a look, I fixed it by moving it into nested if (but I know you hate them 😅)

Reply:
ok... whatever, don't take my advice